nexus/src/db/collection_insert.rs

```diff
@@ -21,76 +21,15 @@ use diesel::query_builder::*;
 use diesel::query_dsl::methods as query_methods;
 use diesel::query_source::Table;
 use diesel::sql_types::SingleValue;
+use nexus_db_model::DatastoreCollection;
 use std::fmt::Debug;
 use std::marker::PhantomData;
 
-/// Trait to be implemented by any structs representing a collection.
-/// For example, since Organizations have a one-to-many relationship with
-/// Projects, the Organization datatype should implement this trait.
-/// ```
-/// # use diesel::prelude::*;
-/// # use omicron_nexus::db::collection_insert::DatastoreCollection;
-/// # use omicron_nexus::db::model::Generation;
-/// #
-/// # table! {
-/// #     test_schema.organization (id) {
-/// #         id -> Uuid,
-/// #         time_deleted -> Nullable<Timestamptz>,
-/// #         rcgen -> Int8,
-/// #     }
-/// # }
-/// #
-/// # table! {
-/// #     test_schema.project (id) {
-/// #         id -> Uuid,
-/// #         time_deleted -> Nullable<Timestamptz>,
-/// #         organization_id -> Uuid,
-/// #     }
-/// # }
-///
-/// #[derive(Queryable, Insertable, Debug, Selectable)]
-/// #[diesel(table_name = project)]
-/// struct Project {
-///     pub id: uuid::Uuid,
-///     pub time_deleted: Option<chrono::DateTime<chrono::Utc>>,
-///     pub organization_id: uuid::Uuid,
-/// }
-///
-/// #[derive(Queryable, Insertable, Debug, Selectable)]
-/// #[diesel(table_name = organization)]
-/// struct Organization {
-///     pub id: uuid::Uuid,
-///     pub time_deleted: Option<chrono::DateTime<chrono::Utc>>,
-///     pub rcgen: Generation,
-/// }
-///
-/// impl DatastoreCollection<Project> for Organization {
-///     // Type of Organization::identity::id and Project::organization_id
-///     type CollectionId = uuid::Uuid;
-///
-///     type GenerationNumberColumn = organization::dsl::rcgen;
-///     type CollectionTimeDeletedColumn = organization::dsl::time_deleted;
-///
-///     type CollectionIdColumn = project::dsl::organization_id;
-/// }
-/// ```
-pub trait DatastoreCollection<ResourceType> {
-    /// The Rust type of the collection id (typically Uuid for us)
-    type CollectionId: Copy + Debug;
-
-    /// The column in the CollectionTable that acts as a generation number.
-    /// This is the "child-resource-generation-number" in RFD 192.
-    type GenerationNumberColumn: Column + Default;
-
-    /// The time deleted column in the CollectionTable
-    // We enforce that this column comes from the same table as
-    // GenerationNumberColumn when defining insert_resource() below.
-    type CollectionTimeDeletedColumn: Column + Default;
-
-    /// The column in the ResourceTable that acts as a foreign key into
-    /// the CollectionTable
-    type CollectionIdColumn: Column;
-
+/// Extension trait adding behavior to types that implement
+/// `nexus_db_model::DatastoreCollection`.
+pub trait DatastoreCollectionExt<ResourceType>:
```

Comment:
It seems like:

- there are two abstractions here: a `DatastoreCollection` and a `DatastoreAttachTarget`
- some part of those traits moved into `db::model` because they're basically part of the model now
- other parts of those traits stayed in `db`, which means we grew the `*Ext` traits

I'm not that clear on the relationship between `DatastoreCollection` and `DatastoreAttachTarget` (which looks like a particular kind of collection) but I think that predates this change.  I wonder whether it would be clearer to rename `DatastoreCollection` -> `DatastoreCollectionConfig` and `DatastoreCollectionExt` -> `DatastoreCollection`.  I say that because I don't think there's any point to a raw `DatastoreCollection` as-is right now -- the point is what's provided by `DatastoreCollectionExt`.  It may be that I'm misremembering what this subsystem does.

Reply:
> It seems like:
> 
> * there are two abstractions here: a `DatastoreCollection` and a `DatastoreAttachTarget`
> * some part of those traits moved into `db::model` because they're basically part of the model now
> * other parts of those traits stayed in `db`, which means we grew the `*Ext` traits

All correct, with the note that there are two more traits that stayed fully in `nexus` (`DatastoreDetachTarget` and `DatastoreDetachManyTarget`), because they themselves are implemented as extensions of `DatastoreAttachTarget`.

> I wonder whether it would be clearer to rename `DatastoreCollection` -> `DatastoreCollectionConfig` and `DatastoreCollectionExt` -> `DatastoreCollection`

I like this; I'll do the same for `DatastoreAttachTarget`.

Reply:
Trait names changed in 330f86e6.